src/xstate/atomWithMachine.ts

```diff
@@ -154,9 +156,21 @@ export function atomWithMachine<
 
   const machineStateWithServiceAtom = atom(
     (get) => get(machineStateAtom),
-    (get, _set, event: Parameters<AnyInterpreter['send']>[0]) => {
+    (
+      get,
+      set,
+      event: Parameters<AnyInterpreter['send']>[0] | typeof RESTART
+    ) => {
       const { service } = get(machineAtom)
-      service.send(event)
+      if (event === RESTART) {
+        service.stop()
+        set(cachedMachineAtom, null)
+        set(machineAtom, null)
+        set(machineStateAtom, () => void 0)
```

Comment:
This doesn't seem correct.
we can't stop machine on unmount.

A possible fix would be to make L133 `service.stop()` to
```js
  const { service } = get(machineAtom)
  service.stop()
```
and change here:
```suggestion
        const { service: newService } = get(machineAtom)
        newService.onTransition((nextState: any) => {
          set(cachedMachineStateAtom, nextState)
        })
        newService.start()
```
and remove L170.

I'm not so sure...

Reply:
the proposal LGTM